File: app/src/net/reichholf/dreamdroid/parsers/enigma2/saxhandler/E2SimpleListHandler.java
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 * 
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.parsers.enigma2.saxhandler;

import org.xml.sax.Attributes;
import org.xml.sax.helpers.DefaultHandler;

import java.util.ArrayList;

public abstract class E2SimpleListHandler extends DefaultHandler {
	
	private boolean inItem;
	
	private String mTag;
	private String mItem;
	
	protected ArrayList<String> mList;
	
	protected E2SimpleListHandler(String tag){
		mTag = tag;
	}
	
	public void setList(ArrayList<String> list){
		mList = list;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see org.xml.sax.helpers.DefaultHandler#startElement(java.lang.String,
	 * java.lang.String, java.lang.String, org.xml.sax.Attributes)
	 */
	@Override
	public void startElement(String namespaceUri, String localName, String qName, Attributes attrs) {
		if (localName.equals(mTag)) {
			inItem = true;
			mItem = "";
		}
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see org.xml.sax.helpers.DefaultHandler#endElement(java.lang.String,
	 * java.lang.String, java.lang.String)
	 */
	@Override
	public void endElement(String namespaceURI, String localName, String qName) {
		if (localName.equals(mTag)) {
			inItem = false;
			mList.add(mItem.trim());
		}
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see org.xml.sax.helpers.DefaultHandler#characters(char[], int, int)
	 */
	@Override
	public void characters(char ch[], int start, int length) {
		String value = new String(ch, start, length);
		if (inItem) {
			mItem += value;
		}
	}
}

File: app/src/net/reichholf/dreamdroid/parsers/enigma2/saxhandler/E2SimpleListHandler.kt
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 *
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.parsers.enigma2.saxhandler

import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler

abstract class E2SimpleListHandler(private val mTag: String?) : DefaultHandler() {
    private var inItem: Boolean = false
    private var mItem: String? = null

    @JvmField
    protected var mList: ArrayList<String>? = null

    fun setList(list: ArrayList<String>?) {
        mList = list
    }

    override fun startElement(
        namespaceUri: String?,
        localName: String,
        qName: String?,
        attrs: Attributes?,
    ) {
        if (localName == mTag) {
            inItem = true
            mItem = ""
        }
    }

    override fun endElement(namespaceURI: String?, localName: String, qName: String?) {
        if (localName == mTag) {
            inItem = false
            mList!!.add(mItem!!.trim())
        }
    }

    override fun characters(ch: CharArray, start: Int, length: Int) {
        val value = String(ch, start, length)
        if (inItem) {
            mItem += value
        }
    }
}
